Disabled `NaturalHomomorphism` for `FactorGroup`, as it has side effects.

hulpke opened the pull request:
(Moved into obsoletes).
This resolves #4656

Also included the (trivial) fix for #4698

## Text for release notes

The use of `FactorGroup` succeeded by `NaturalHomomorphism` has been made obsolete, as the implementation relied on side-effects.

fingolfin commented on lib/grp.gi:
```diff
@@ -2685,6 +2685,9 @@ end );
     [ IsGroup ],
 function(G)
+  Error("The use of `NaturalHomomorphism` for a `FactorGroup`\n",
+    "has been removed, as it caused side-effects.");
```
But won't that break existing user code that relies on this?

hulpke replied:
Yes, should code rely onthis functionality. We could change it to add text "enter `return;` to continue, or change the error to a warning message.

fingolfin replied:
Also we should perhaps quickly check if any distributed packages rely on this?

hulpke replied:
I have checked through the packages in 4.11. Only XMod,  laguna, primgrp, rds, wedderga and xgap call `FactorGroup(`. None of them uses `NaturalHomomorphism(`. I will change the error to a warning.

cdwensley replied:
XMod does use NaturalHomomorphism

hulpke replied:
@cdwensley Where does it use `NaturalHomomorphism`? I found it uses `NaturalHomomorphismByNormalSubgroup` (which is safe and well) only in version 2.82.

cdwensley replied:
I'm happy to change it if that is the sensible thing to do. 
Christopher-Wensleys-iMac:lib christopherwensley$ grep -n NaturalHomomorphism *
catone.gi:575:                nat:=NaturalHomomorphismByNormalSubgroup(G,N);
gp2map.gi:1764:    nat1 := NaturalHomomorphism( KR1 ); 
gp2map.gi:1795:                          nat2 := NaturalHomomorphism( KQ ); 
gp2map.gi:1862:    nat1 := NaturalHomomorphism( KR1 ); 
gp2map.gi:1891:                        natKQ := NaturalHomomorphism( KQ );
gp2map.gi:1897:                        natKR1 := NaturalHomomorphism( KR1 );
gp2map.gi:1902:                        natKQ := NaturalHomomorphism( KQ );

cdwensley replied:
Yes, this does appear to be a recent change.

hulpke replied:
Since this is a recent change (and thus possibly not even in a distributed version) I would urge to replace code of the form
```
F:=FactorGroup(G,N);
hom:=NaturalHomomorphism(F);
```
by
```
hom:=NaturalHomomorphismByNormalSubgroup(G,N);
F:=Image(hom,G);
```
The cost will be the same, but it does not rely on the side-effect of aching the homomorphism in the object.

cdwensley replied:
It all happened when quasi-isomorphisms were introduced: PR #107 (17/2/21). 
I'll try using NaturalHomomorphismByNormalSubgroup instead.

fingolfin commented on lib/obsolete.gi:
```diff
@@ -1029,6 +1029,21 @@ end );
+    [ IsGroup ],
+function(G)
+  Info(InfoWarning,0,"The use of `NaturalHomomorphism` for a `FactorGroup`\n",
+    "has been removed, as it caused side-effects.\n",
```
It has not (yet) been removed, though, has it? Perhaps say "deprecated" or so instead?
```suggestion
    "has been deprecated, as it caused side-effects.\n",
```

fingolfin reviewed (approved):
Thank you

cdwensley commented:
I recall now that this advice was given and followed some years ago.
As the operation was still available, it was used again this year.
The 5 occurrences of NaturalHomomorphism in XMod have now been replaced.
Since this is a recent change (and thus possibly not even in a 
distributed version) I would urge to replace code of the form
    F:=FactorGroup(G,N);   hom:=NaturalHomomorphism(F);
by
    hom:=NaturalHomomorphismByNormalSubgroup(G,N);
    F:=Image(hom,G);
The cost will be the same, but it does not rely on the side-effect of 
aching the homomorphism in the object.

fingolfin commented:
Unfortunately something in here broke the test suite, specifically `tst/teststandard/grpprmcs.tst:1813`, which I didn't notice as there is this 32bit float CI error... *sigh*. 
```
testing: /home/runner/work/gap/gap/tst/teststandard/grpprmcs.tst
########> Diff in /home/runner/work/gap/gap/tst/teststandard/grpprmcs.tst:1813
# Input is:
ImagesRepresentative(iso,Product(GeneratorsOfGroup(g)));
# Expected output:
F3*F4*F5*F2*F1^-1*F2^2*(F1*F2^-1)^2
# But found:
F3*F4*F5*(F1^-1*F2^-1)^2
########
```

hulpke commented:
Yes. This is a side-effect of a small improvement to `SmallGeneratingSet' which I had added since the PR was hanging anyhow (did not realize you would then override the merge).
The test is simply for evaluating the homomorphism on an element that is not a generator, but since the generating set changed, the image (entirely harmlessly) also changed. Do you want me to submit a PR for changing the test example?
